### DISCUSSION TITLE
Using a 2nd machine/zwave js instance for inclusion/exclusion

### DISCUSSION THREAD

**Discussion by rlippmann** [zwave-js/zwave-js-ui#3879] [Q&A]
Hi all,

So, I have a Z-stick 5 on my server.  Running zwave-js-ui, home assistant.  Everything works fine.

The problem is sometimes I have issues including/excluding because of distance/quirkiness of certain z-wave devices.

What I'd like to do is use the z-stick on a laptop so I can just move the controller near the device.

I'm wondering:
* is this possible
* how to go about doing this

I assume that since the controller holds the node information, this should be possible.  So the procedure should be:

* install zwave js on second machine
* copy over S0/S2 codes

Then when I need to use the controller for a short range inclusion I should be able to plug it into my laptop, do the inclusion, and put the controller back on the main server.

Does this sound correct?  Do I need to back up/restore nvram for this?  Am I missing something?

Thanks...

**Comment by kpine** [comment]
Z-Wave JS supports inclusion controllers, so if you have another controller and other software you could include it as a secondary controller and perform inclusions from it. Even Silicon Labs' PC Controller software (Windows only) should work for that. However, Z-Wave JS doesn't support itself being a secondary controller, so you can't use it for that purpose, yet. https://zwave-js.github.io/node-zwave-js/#/getting-started/faq?id=does-z-wave-js-support-secondary-controllers:

> Z-Wave JS expects to be the primary controller in the network and it will try to assume that role when started. It will not work correctly as a secondary controller.

It looks like being a secondary controller is currently being implemented though for a future release. https://github.com/zwave-js/node-zwave-js/pull/7135. Of course, ZUI will also have to update to support it.

> Then when I need to use the controller for a short range inclusion I should be able to plug it into my laptop, do the inclusion, and put the controller back on the main server.

This would work (I would just copy the entire store directory to preserve all settings and driver cache), but moving the controller is generally a bad idea. Your USB controller is a static device (expected to stay in a permanent position). When nodes are included routes are calculated based on the current network environment. If you move the controller back, communication with the included node will probably fail, and possibly other nodes might as well if routes were recalculated in the other position. If there is no other choice, preference would be to move the device close to the controller and then moved back, after doing rebuilds. Moving the controller would be the very last choice. <https://zwave-js.github.io/node-zwave-js/#/troubleshooting/common-issues?id=inclusion-finds-no-new-devices>

**Comment by rlippmann** [comment]
Sorry, I guess I wasn't clear.  I have 1 controller.  I want to attach it to another machine to perform inclusion, then put it back.

**Reply by kpine** [reply]
No, you were clear. That first part of my reply was mostly regarding:

> The problem is sometimes I have issues including/excluding because of distance/quirkiness of certain z-wave devices.

Having a second inclusion controller would solve that, so you wouldn't have to move the primary around.

**Comment by rlippmann** [comment]
Actually, now that I'm thinking about it, I wonder if USB/IP might work....

**Reply by kpine** [reply]
Placing the primary controller in a central location is preferred. If you are able to achieve that using TCP serial ports, than that's an option and it's supported by Z-Wave JS UI. https://zwave-js.github.io/node-zwave-js/#/usage/tcp-connection

A similar approach would be to put ZUI on an SBC like a Raspberry Pi.

Also, before you do anyone of those, make sure you're using a USB extension cable. If your setup involves an equipment rack, the stick should be outside of it.

https://zwave-js.github.io/node-zwave-js/#/troubleshooting/first-steps?id=controller-placement

**Comment by rlippmann** [comment]
So, I just remembered that since I have an Aeotec Z-Stick I can just remove it from the USB port and move it around to pair.

I assume Z-wave JS UI will pick up the new nodes when I reconnect?

**Reply by kpine** [reply]
I'd say this method is much worse than using a laptop, except for the convenience. First problem, you have to re-interview the devices when moved back to Z-Wave JS UI. Battery devices will need to be woken up manually, wired devices are automatic. If the routes are messed up, communication may not work. Second problem, it doesn't support S0 or S2 Security at all, so you can only include with no security.

I couldn't find this same information in the docs, but this issue describes the problems of moving the controller around: https://github.com/zwave-js/node-zwave-js/issues/1995.

> This method is a recipe for disaster and reliably causes problems.
